Improve initial scale calculation

jgmdev opened the pull request:
This PR modifies the get_scale() C function used on main to calculate the initical scale by using a base width and height which works better than current strategy based on the dpi value returned by sdl, 1280x720 was chosen as base because it is legible with the default font size of 15, here a comparison of before and after on differently tested resolutions:

1400x900 before
![1400x900-by-dpi](https://user-images.githubusercontent.com/1702572/170557447-ba740b92-b733-40bb-bbac-3533bd0f5d17.png)

1400x900 after
![1400x900-by-bw](https://user-images.githubusercontent.com/1702572/170557611-5c90e63f-9d4f-40f4-ba3c-61c5b66019ce.png)

1600x900 before
![1600x900-by-dpi](https://user-images.githubusercontent.com/1702572/170557673-f70e23ba-fc63-469d-af47-e3fa5ec056b7.png)

1600x900 after
![1600x900-by-bw](https://user-images.githubusercontent.com/1702572/170557776-0816e63f-bf37-4f62-aaa4-a39d0146fb9a.png)

1920x1080 before
![1920x1080-by-dpi](https://user-images.githubusercontent.com/1702572/170557844-884d90ff-364a-4400-a816-fa13f2d12e0e.png)

1920x1080 after
![1920x1080-by-bw](https://user-images.githubusercontent.com/1702572/170557887-ac353770-d1cd-4795-8e00-8e92bbdb1233.png)

adamharrison commented:
Yeah; this calculation has always been a little messed up. I think this something like this is the way to go.

jgmdev commented:
I wonder if dropping the dpi fall back and also allow the base resolution calculation for mac too

jgmdev commented:
Reminder note: non sdl renderer gets broken with float Rect and the [improve unstable text y axis on fractional scales](https://github.com/lite-xl/lite-xl/pull/998/commits/cb57b5b1729b493b2f5cd09a81b3eda6c8203a0f) commit which needs to be reverted for it to work and something like the following to get both SDL Renderer and non sdl renderer working.

```diff
+#ifdef LITE_USE_SDL_RENDERER
+#define RECT_TYPE float
+#else
+#define RECT_TYPE int
+#endif
+
 #define FONT_FALLBACK_MAX 4
 typedef struct RenFont RenFont;
 typedef enum { FONT_HINTING_NONE, FONT_HINTING_SLIGHT, FONT_HINTING_FULL } ERenFontHinting;
 typedef enum { FONT_ANTIALIASING_NONE, FONT_ANTIALIASING_GRAYSCALE, FONT_ANTIALIASING_SUBPIXEL } ERenFontAntialiasing;
 typedef enum { FONT_STYLE_BOLD = 1, FONT_STYLE_ITALIC = 2, FONT_STYLE_UNDERLINE = 4 } ERenFontStyle;
 typedef struct { uint8_t b, g, r, a; } RenColor;
-typedef struct { float x, y, width, height; } RenRect;
+typedef struct { RECT_TYPE x, y, width, height; } RenRect;
```

jgmdev commented:
Reminder note:  resizing the window fetches new scale which most of the time differs by a minimal change in decimal points when doing fractional scaling, but this is enough to cause jumpy text each time the window is resized, a change as follows only applies a new scale if different enough which eliminates the jumpy text at least on window resizing:

```c
static float scale_x = 0, scale_y = 0;
static void update_surface_scale(RenWindow *ren) {
  /* cache the scale to prevent jumpy text on resizing */
  if (scale_x <= 0) {
    int w_pixels, h_pixels;
    int w_points, h_points;
    SDL_GL_GetDrawableSize(ren->window, &w_pixels, &h_pixels);
    SDL_GetWindowSize(ren->window, &w_points, &h_points);
    float temp_scale_x = (float) w_pixels / (float) w_points;
    float temp_scale_y = (float) h_pixels / (float) h_points;
    if (roundf(temp_scale_x * 100) / 100 != roundf(scale_x * 100) / 100) {
      scale_x = temp_scale_x;
      scale_y = temp_scale_y;
    }
  }
  ren->surface_scale_x = scale_x;
  ren->surface_scale_y = scale_y;
}
```

jgmdev commented:
Removed the `improve unstable text y axis on fractional scales` and added other minor changes for more pristine renderer near to what we had before which still works with non sdl renderer. So next step is to properly trouble shoot the root cause of the jumpy text on y axis when a fractional scale is been used on window resize and scroll.

Jan200101 commented:
This works for me but I wondered why so I did some digging:
the monitor DPI and the X DPI are not nesserarily the same.

So, on my laptop with an active resolution of 1920px x 1080px and display dimensions of 344mm x 194mm (values can be gathered using `xrandr | grep -w connected`.
This comes out to a rough DPI of 144, which is what SDL returns my DPI to be, resulting in a Scale of `144 / 96 = 1.5` being used.
My actual X DPI is 96, this results in the Lite-XL window being scaled larger than intended.

I am unsure if the solution in this PR is correct, but it resolves my issue since it no longer asks for the display DPI but for the X DPI.
